[T3A3][F12-A3]Joel Lim Jing

aljorhythm opened the pull request:
Ready for review

aljorhythm committed:
Hide feedbackToUser attribute in class seedu.addressbook.commands.CommandResult
Generate getter for feedbackToUser attribute in class seedu.addressbook.commands.CommandResult

aljorhythm committed:
Create class seedu.addressbook.data.person.Street
Create class seedu.addressbook.data.person.Unit
Create class seedu.addressbook.data.person.Block
Create class seedu.addressbook.data.person.PostalCode

aljorhythm committed:
Use newly created classes Block, Unit, PostalCode, Street in seedu/addressbook/data/person/Address.java

aljorhythm committed:
Add comments and correct coding style for Block.java, PostalCode.java, Street.java, Unit.java

funnieguy commented:
Work in progress

aljorhythm committed:
Fix bug Address::value has been removed

aljorhythm committed:
Fix bug Address:value is not accessible anymore

okkhoy commented on src/seedu/addressbook/commands/CommandResult.java:
```diff
@@ -11,7 +11,7 @@
 
     /** The feedback message to be shown to the user. Contains a description of the execution result */
-    public final String feedbackToUser;
+    private final String feedbackToUser;
```
this commit doesn't belong to T3A3.
Avoid accumulating unrelated changes into a single PR. 
Tip: If you are currently on `branch1` and want to create a new `branch2`, you need to switch back to `master` branch before creating `branch2`. If not, changes in `branch1` will appear in the PR created from `branch2` later.

okkhoy commented on src/seedu/addressbook/data/person/Address.java:
```diff
@@ -3,30 +3,38 @@
         }
-        this.value = trimmedAddress;
+        String[] addressParts = trimmedAddress.split(", ");
+        this.block = new Block(addressParts[0]);
```
why not 1, or 2 or 100?
I hope you understand what I mean here.

okkhoy commented on src/seedu/addressbook/data/person/Address.java:
```diff
@@ -3,30 +3,38 @@
+        String[] addressParts = trimmedAddress.split(", ");
+        this.block = new Block(addressParts[0]);
+        this.street = new Street(addressParts[1]);
+        this.unit = new Unit(addressParts.length > 2 ? addressParts[2] : "");
```
this is kinda good thing to have. you are providing default null values for non supplied fields.

okkhoy commented on src/seedu/addressbook/data/person/Address.java:
```diff
@@ -38,19 +46,31 @@ public static boolean isValidAddress(String test) {
     @Override
     public String toString() {
-        return value;
+        String[] addressParts = new String[] { block.toString(), street.toString(), unit.toString(), postalCode.toString() };
```
do you notice the code here has a neat arrow head shape? it is generally an indication of SLAP violation. see if you can do better than this here.

okkhoy commented on src/seedu/addressbook/data/person/Address.java:
```diff
@@ -38,19 +46,31 @@ public static boolean isValidAddress(String test) {
-        return value;
+        String[] addressParts = new String[] { block.toString(), street.toString(), unit.toString(), postalCode.toString() };
+        String ret = "";
+        for(int i = 0; i < addressParts.length; i++){
```
need space betwee `){` all over this block of code.

okkhoy commented on src/seedu/addressbook/data/person/Block.java:
```diff
@@ -0,0 +1,18 @@
+/**
+ * 
+ * @author joellim
+ *  The block part of an address
```
comments should be in active verb form.
`Represents the block information in an address` may be more appropriate

okkhoy commented on src/seedu/addressbook/data/person/PostalCode.java:
```diff
@@ -0,0 +1,19 @@
+
+/**
+ * 
+ * @author joellim
```
you can possibly remove the `@author` tag; your git commits identify you any way :-)

okkhoy reviewed (commented):
please note that you need  to update the documentation to reflect these changes you made to the code (in program help and user guide).
also, you need to update the test files to test the new behavior of splitting address to components!

okkhoy commented:
@funnieguy some comments added, please take a look. please close the PR once you have noted the suggestions.

aljorhythm committed:
Replace magic numbers for indices of address parts in address string with constants (AddressCSVIndices)
Change Address.Example to reflect human-readable legal syntax for parsable string
Rename variable trimmedAddress in Address(String, boolean) to address

Changed Address.toString() to simply join all parts of address with “, “

Changed Address validation regex to accept string delimited by “, “

Changed test cases and expected results

aljorhythm committed:
Changed address example in AddCommand.java

aljorhythm committed:
Added try/catch block to initiation of address part instances to catch invalid formatting
Change constructor Block(String) to throw IllegalValueException if format of block is illegal
Added test case inputs
Changed expected output

aljorhythm committed:
Created IllegalValueException.matchString(String, String, String) to abstract matching regex and throwing of IllegalValueException in address part constructors
Change blocked to reflect abstraction
Added regex testing for class PostalCode
Added regex testing for class Street

funnieguy commented:
I commented "Work in progress" after Friday's tutorial because I was in the process of making changes. I realise I should have closed the pull request and issue another one. I have made more commits since then.

okkhoy commented on src/seedu/addressbook/data/person/Address.java:
```diff
@@ -44,28 +48,22 @@ public static boolean isValidAddress(String test) {
                 || (other instanceof Address // instanceof handles nulls
                         && this.toString().equals(((Address) other).toString())); // state
-                                                                                  // check
+        // check
```
unnecessary comment here

aljorhythm replied:
this is due to the automatic formatting in Eclipse, will disable comment formatting

okkhoy commented on src/seedu/addressbook/data/person/Address.java:
```diff
@@ -8,33 +8,37 @@
     private Unit unit;
     private PostalCode postalCode;
 
+    private static class AddressCSVIndices {
```
this name is a bit iffy.
i get where you are coming from, but seems like you are introducing something complex.
try to think of a better name, if possible?

okkhoy commented on src/seedu/addressbook/data/exception/IllegalValueException.java:
```diff
@@ -10,4 +10,20 @@
     }
+
+    /**
+     * 
```
you need to tell what this method does at least.

aljorhythm replied:
Roger

aljorhythm replied:
note taken

okkhoy commented on src/seedu/addressbook/data/person/Address.java:
```diff
@@ -35,10 +35,15 @@ public Address(String address, boolean isPrivate) throws IllegalValueException {
-        this.street = new Street(addressParts.length > 1 ? addressParts[AddressCSVIndices.STREET] : "");
-        this.unit = new Unit(addressParts.length > 2 ? addressParts[AddressCSVIndices.UNIT] : "");
-        this.postalCode = new PostalCode(addressParts.length > 3 ? addressParts[AddressCSVIndices.POSTAL_CODE] : "");
+        try {
```
what if there are no commas at all?

aljorhythm replied:
Actually the regex "^[^,]*(, [^,]*){3}$" checks for exactly 3 commas so the checking of length is made redundant. My test cases have been updated to check for an error if address is not "<something>, [something], [something], [something]"

okkhoy commented on src/seedu/addressbook/data/person/Block.java:
```diff
@@ -8,8 +10,11 @@
 
-    public Block(String block) {
-        this.block = block;
+    public static String REGEX = "(^[\\d]{1,4}[A-Z]{0,1}$)|(^$)";
```
i m not sure what this regex is doing! i got confused trying to work through :-P
may be this warrants a comment in the code as to what it does (you are the better judge here)

aljorhythm replied:
(^[\\d]{1,4}[A-Z]{0,1}$) - begin with 1 to 4 digits followed by optional capital letter
^$ - empty
Block can be empty because not all addresses have block numbers
I did throw a message in the constructor if the regex matching fails which documents this but a better way is to have a ERROR_MESSAGE class variable on top of REGEX

okkhoy reviewed (commented):
sorry! i assumed you forgot about the PR after you added WIP! i will take note next time.

I have reviewed your changes now. please take a look

okkhoy commented:
@funnieguy i have added some more comments.

aljorhythm commented:
Thanks @okkhoy
